dyson.js

```diff
@@ -24,144 +24,262 @@ var mathjs = require('mathjs');
+                    cube.matrixAutoUpdate = false;
+                    this.grid.add(cube);
+                    xdim.push(cube);
+                }
+                zdim.push(xdim);
+            }
+            meshCells.push(zdim);
+        }
+    }
+
+    //this.grid.rotation.z = -45 * mathjs.PI / 180;
+    //this.grid.rotation.x = -60 * mathjs.PI / 180;
+
+    //this.grid.rotation.x = -60 * mathjs.PI / 180;
+    this.grid.position.x = -30;
+    //this.grid.position.y = -100;
+    //this.grid.position.z = -100;
 
-    light = new THREE.PointLight(0xFFFFFF);
+    scene.add(this.grid);
+};
+
+/*
+ * Called once to render the canvas into the DOM with the initial cell data.
+ */
+HtmCellVisualization.prototype.render = function() {
+    var me = this;
+    var renderer = this.renderer;
+    var scene = this.scene;
+    var controls = this.controls;
+    var camera = this.camera;
+    var light = this.light;
+    var w = this.width;
+    var h = this.height;
 
-    scene.add(light);
+    this._createMeshCells();
 
     renderer = new THREE.WebGLRenderer( { antialias: false } );
 
     renderer.setClearColor( scene.fog.color );
     renderer.setPixelRatio( window.devicePixelRatio );
-    renderer.setSize( innerWidth, innerHeight );
+    renderer.setSize(w, h);
 
-    if (elementId) {
-        document.getElementById(elementId).appendChild(renderer.domElement);
-    } else {
-        document.body.appendChild(renderer.domElement);
+    function innerRender() {
+        light.position.x = camera.position.x;
+        light.position.y = camera.position.y;
+        light.position.z = camera.position.z;
+        renderer.render(scene, camera);
     }
 
+    this.controls.addEventListener('change', innerRender);
+
     window.addEventListener('resize', function() {
-        camera.aspect = innerWidth / innerHeight;
+        w = me.width = me.$container.innerWidth();
+        h = me.height = me.$container.innerHeight();
+        camera.aspect = w / h;
         camera.updateProjectionMatrix();
-        renderer.setSize( innerWidth, innerHeight );
+        renderer.setSize(w, h);
         controls.handleResize();
-        render();
+        innerRender();
     }, false );
 
-    render();
+    this.$container.append(renderer.domElement);
 
-    this.camera = camera;
-    this.renderer = renderer;
-    this.scene = scene;
-    this.controls = controls;
-    this.light = light;
-    this.animate = animate;
-}
+    function animate() {
+        requestAnimationFrame(animate);
+        me.controls.update();
+        innerRender();
+    }
 
+    setTimeout(animate, 0);
 
-function animate() {
-    requestAnimationFrame( animate );
-    controls.update();
-}
+    camera.position.z = 20;
+};
 
-function render() {
-    light.position.x = camera.position.x;
-    light.position.y = camera.position.y;
-    light.position.z = camera.position.z;
-    renderer.render( scene, camera );
-}
+HtmCellVisualization.prototype.redraw = function() {
+    this._applyMeshCells();
```

Comment:
This is the expensive call, so it is only called on first render and whenever the client calls redraw. If we put `_applyMeshCells` inside the animation loop, it grinds to a halt.